src/devices/Card/CreditCard/Tag.cs

```diff
@@ -0,0 +1,116 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Iot.Device.Card.CreditCardProcessing
+{
+    /// <summary>
+    /// A Tag class containing part of a card information
+    /// </summary>
+    public class Tag
```

Comment:
If this is about ASN.1/BER, it should be internal, and/or have a more specific name.

A BER/CER/DER reader/writer should be in the next version of .NET Core, with the current prototype at https://github.com/dotnet/corefxlab/tree/master/src/System.Security.Cryptography.Asn1.Experimental / https://dotnet.myget.org/feed/dotnet-corefxlab/package/nuget/System.Security.Cryptography.Asn1.Experimental

Reply:
So I'll use it as soon as it will be release. I had a look at it and seems like it can fully work. @krwq , should I use it even if in preview?

Reply:
@Ellerbach we should probably wait a bit till it is shipped as stable. @bartonjs is the plan to move that to corefx? Will it ship as a OOB package? We currently target nc2.1 with our bindings package and it looks like the current implementation you have does as well, I suppose we intend to keep it that way once we move it to corefx?

Reply:
> is the plan to move that to corefx?

Actually, it's already in corefx and in use, but as internal.

> Will it ship as a OOB package?

My expectation is that it'll be part of System.Security.Cryptography.Encoding, inbox-only.  But maybe we have enough need that it'll be half-in half-out.

Reply:
So I'll evaluate how it can be used and if it fills the need. All objects are not with standard tags, they are all custom ones, so if the decoder is not supporting this, I won't be able to use it..

Reply:
Aside from ASN.1's Real type, that library should do pretty much everything.  Even with Real it'll understand the tag and length, let you get the payload uninterpreted, and then move to the next thing.

Reply:
I've been playing a bit with it. It's missing functions to get just raw data and raw tag numbers. It's very much oriented to extract know types and not much for raw.
That said, all the internal functions are present but not exposed and the way the tag is managed in the code makes it not possible to get the raw tag number.

Simple example below (those are real data that I read from cards):

```csharp
            byte[] span = new byte[] { 0x6F, 0x69, 0x84, 0x0E, 0x32, 0x50, 0x41, 0x59, 0x2E, 0x53, 0x59, 0x53, 0x2E, 0x44, 0x44, 0x46, 0x30, 0x31, 0xA5, 0x57, 0xBF, 0x0C, 0x54, 0x61, 0x28, 0x4F, 0x07, 0xA0, 0x00, 0x00, 0x00, 0x42, 0x10, 0x10, 0x50, 0x0B, 0x43, 0x42, 0x20, 0x43, 0x4F, 0x4D, 0x50, 0x54, 0x41, 0x4E, 0x54, 0x9F, 0x2A, 0x01, 0x03, 0x87, 0x01, 0x01, 0x9F, 0x0A, 0x08, 0x00, 0x01, 0x05, 0x02, 0x00, 0x00, 0x00, 0x00, 0x61, 0x28, 0x4F };
            AsnReader reader = new AsnReader(span, AsnEncodingRules.BER);
            // This gives the tag number 0x0F while
            // Real raw tag number is 0x6F
            var ret = reader.PeekTag();
            // this break content is seen as non valid!
            var data = reader.PeekEncodedValue();
            // This break content is seen as non valid!
            var content = reader.PeekContentBytes();
```

As commented in the code, the raw tag number should be 0x6F, it is Tagged 0x0F. Reason is because when decoding the tag, only 7 bits are kept while the raw value needs 8 bits. See here: https://github.com/dotnet/corefxlab/blob/master/src/System.Security.Cryptography.Asn1.Experimental/System/Security/Cryptography/Asn1/Asn1Tag.cs#L210

So once the tag is decoded, it's not possible to get the raw value of the tag.

And because this value is then interpreted as another tag, in this case "reserved" which is not processed, I get an error when I want to extract the data.

So the functiosn are great for known types but not built for raw usage like in SmartCards for example. That would need to be adapted to support raw tag numbers and raw data.

Reply:
I've been playing a bit with it. It's missing functions to get just raw data and raw tag numbers. It's very much oriented to extract know types and not much for raw.
That said, all the internal functions are present but not exposed and the way the tag is managed in the code makes it not possible to get the raw tag number.

Simple example below (those are real data that I read from cards):

```csharp
            byte[] span = new byte[] { 0x6F, 0x69, 0x84, 0x0E, 0x32, 0x50, 0x41, 0x59, 0x2E, 0x53, 0x59, 0x53, 0x2E, 0x44, 0x44, 0x46, 0x30, 0x31, 0xA5, 0x57, 0xBF, 0x0C, 0x54, 0x61, 0x28, 0x4F, 0x07, 0xA0, 0x00, 0x00, 0x00, 0x42, 0x10, 0x10, 0x50, 0x0B, 0x43, 0x42, 0x20, 0x43, 0x4F, 0x4D, 0x50, 0x54, 0x41, 0x4E, 0x54, 0x9F, 0x2A, 0x01, 0x03, 0x87, 0x01, 0x01, 0x9F, 0x0A, 0x08, 0x00, 0x01, 0x05, 0x02, 0x00, 0x00, 0x00, 0x00, 0x61, 0x28, 0x4F };
            AsnReader reader = new AsnReader(span, AsnEncodingRules.BER);
            // This gives the tag number 0x0F while
            // Real raw tag number is 0x6F
            var ret = reader.PeekTag();
            // this break content is seen as non valid!
            var data = reader.PeekEncodedValue();
            // This break content is seen as non valid!
            var content = reader.PeekContentBytes();
```

As commented in the code, the raw tag number should be 0x6F, it is Tagged 0x0F. Reason is because when decoding the tag, only 7 bits are kept while the raw value needs 8 bits. See here: https://github.com/dotnet/corefxlab/blob/master/src/System.Security.Cryptography.Asn1.Experimental/System/Security/Cryptography/Asn1/Asn1Tag.cs#L210

So once the tag is decoded, it's not possible to get the raw value of the tag.

And because this value is then interpreted as another tag, in this case "reserved" which is not processed, I get an error when I want to extract the data.

So the functiosn are great for known types but not built for raw usage like in SmartCards for example. That would need to be adapted to support raw tag numbers and raw data.

Reply:
The tag VALUE is 0x6F.  The tag NUMBER is 0x0F with the Application class (0x40) and a constructed encoding (0x20).

The "raw value" can be multiple bytes, so it only exposes it via the Encode method.   In fact, the left shift you highlighted was about reading a tag that is composed of multiple bytes (and not involved here)

If PeekEncodedValue / PeekContentBytes are saying that the content is invalid, then it's invalid. In your case here, the array named `span` claims to have a payload length of 0x69 bytes, but the entire array is only 68 (decimal) bytes long.  length exceeds payload => invalid.

Reply:
hum… so I am was not awake this morning… Will try all gain. Thanks for the explanations. I had then all wrong. Plus the missing byte in the array.

Reply:
SO it finds the value and it's possible to get the content. Now, what is clearly missing is a way to get this raw value directly from the "VALUE" directly as you call it, so in this case 0x6F. In case of longer value, reconstructing it with the "NUMBER" and the application class would be complicated while it can be stored directly as a property.
I realized as well that when you have multiple tag like this one which are not part of the native types, you can't make the cursor move. Only available functions are the Peek ones. So you'll need to use the PeekEncodedValue to get the buffer, get the size, use your original buffer and use another reader to decode what's next. 
I may have miss Something as I missed already few things here. But that looks quite complicated. I guess a function that would make the cursor move and get everything at once will be great.

Reply:
You can get the encoded tag value back by using Asn1Tag.Encode/TryEncode.  If you need it.

```C#
private static readonly Asn1Tag WhateverThisTagIs = new Asn1Tag(TagClass.Application, 0x1F, isConstructed: true);

...

AsnReader outer = new AsnReader(data, AsnEncodingRules.BER);
AsnReader inner = outer.ReadSequence(WhateverTagThisIs);
// Assuming there's not a value expected to be after this one:
outer.ThrowIfNotEmpty();
...
```

Or, if it's contextual:

```C#
if (reader.PeekTag() == WhateverTagThisIs)
{
    ...
}
else
{
    // Maybe just skip it and move to the next thing?
    // ReadEncodedValue returns the whole next value then moves on.
    // So if the return is ignored, it's Skip.
    reader.ReadEncodedValue();
}

You shouldn't ever have to manually slice the ASN structures, the reader does it all.  Everything in ASN breaks down to a built in thing, just use the overloads that take a tag once you have something other than the Universal class tag for the value.